### ISSUE TITLE
Feature request: anchor a set of commands so that chaining is exclusive to the set

### ISSUE THREAD

**Issue by wolfmanstout** [talonvoice/talon#397]
Note: Ryan already added an experiment in beta ("experiment: anchor-file"), but I'm creating this to centralize discussion.

One use case for this is dictation mode, where it's helpful to separate commands which can be interleaved with dictation from those which cannot.  For example, today in knausj dictation_mode there is a sort of "poor man's" support for a subset of generic_editor.talon and repeater.talon commands  where these are hardcoded as simpler variants: 
https://github.com/knausj85/knausj_talon/blob/master/modes/dictation_mode.talon#L13

The issue with simply including these in dictation_mode is that they could be interleaved with prose and fire frequently. What is most desirable is that these commands can be chained amongst themselves but not mixed with actual dictation commands (e.g. prose or formatting commands).  Today, this could theoretically be implemented using captures on one set of commands or the other (dictation or control), but this would mean foregoing the benefits of .talon files for that set. 

Offering support for this will be a balance between expressiveness and simplicity. Perhaps the simplest implementation is the experiment Ryan already implemented: a new header tag which restricts chaining to commands within that file.  In order to achieve the above use case, the simplest way to add this to the dictation_mode.talon file. A more powerful version would allow files to name an "anchor set" so that this anchoring can be applied across multiple files.  In order to align with typical Talon conventions, it would also make sense for this to be paired with a module declaration in Python of the anchor set name.

From my experiments, the only issue with the current experimental implementation is that it does not yet support anchors within anchored files.  Although it's not very helpful to combine this feature with both ^$ anchors  at the same time, it's perfectly reasonable to use it with one or another on any given command.

**Comment by pokey** [comment]
+1 for "anchor set".  Could be used to make a nice compromise between anchored and unanchored phrases.  I would be interested in having a subset of commands that are allowed terminate a formatter phrase, or a phrase used for searching.

**Comment by wolfmanstout** [comment]
@pokey that's a little different from what I'm proposing with anchor-set. My "anchor set" proposal is only a tiny bit more powerful than the current anchor-file feature, allowing multiple files to be chained together in a single set instead of restricting this feature to one file. "chain set" might be a more intuitive name for this feature --  I was just following the naming that Ryan started with the experimental feature. At this point I'm not sure there's evidence that this helps with known use cases --  I added it to help evaluate multiple options. @rntz  proposed something on the beta channel that's more powerful than what I described here, but Ryan raised concerns that it's "extremely confusing and allows really really explosive graphs". I'll flesh out the idea here a bit for the sake of thoroughness of discussion, since it could handle both use cases described here (yours and mine).

The idea is to have a "capture" statement in the header which creates a capture that OR's all the commands in the file together.  When this capture is matched, it results in an object which can be executed with a new action "execute" that runs the  command (or commands if used with +) that was matched. Here's how this would be used to satisfy the original use case described above:

dictation_commands.talon
```
mode: none  # this will only be used as a capture
capture: dictation_commands
-
<user.prose>: auto_insert(prose)
new line: "\n"
...
```

dictation_mode.talon
```
mode: dictation
-
^<user.dictation_commands>+$: execute(dictation_commands_list)
scratch that: user.clear_last_phrase()
select that: user.select_last_phrase()
...
```

And  here it is how it would  allow formatter termination  (which is  a feature I happened to implement in parallel to you):

formatter_terminators.talon
```
# No need to restrict mode, because these are also global commands.
capture: formatter_terminators
-
space: key(space)
enter: key(enter)
...
```

formatters.talon
```
phrase <user.text> <user.formatter_terminators>:
    user.insert_formatted(text, "NOOP")
    execute(formatter_terminators)
...
```

 I'm not sure what you mean by "a phrase used for searching",  perhaps you could clarify?

**Comment by wolfmanstout** [comment]
I've happened upon another use case where something more powerful (e.g. the above "capture" statement) would help, and it looks like @pokey and others might benefit from the same. I'm finding that I can't comfortably use the [prose formatter commands such as "sentence"](https://github.com/knausj85/knausj_talon/blob/master/misc/formatters.talon#L4) because they don't give me a way to capitalize words mid-sentence (i.e. no "cap" unlike the full prose mode).  Since capitalization is fairly unpredictable with Conformer,  I'm getting into the habit of simply using "cap"  whenever I need it other than the beginning of sentences. Hence, I'm about to add support for "cap" to the prose capture. But this means I'm duplicating the equivalent command in the [dictation grammar](https://github.com/knausj85/knausj_talon/blob/master/modes/dictation_mode.talon#L9) with an entirely different implementation. No big deal for this one-off, but it occurs to me that I may end up doing the same with other features in the dictation grammar. In fact, it would be nice to have a way use all the features of dictation mode within a single prefixed utterance, as a way to avoid mode switching for quick prose utterances. If I end up moving all these into a capture, the original benefits of anchor-file are lost -- I might as well use my capture in dictation mode. 

In @pokey's grammar  I see evidence of a similar desire, with a [`prose <user.prose> halt` command](https://github.com/pokey/pokey_talon/blob/main/misc/standard.talon#L58) that currently behaves subtly differently from `prose <user.prose>` followed by `halt` because it doesn't support the commands in dictation mode (cap, newline, etc). Following on my previous comment, here's what this would look like with the "capture" statement available, expanding on the first example with dictation_mode:

standard.talon
```
...
prose <user.dictation_commands>+ halt: execute(dictation_commands_list)
...
```

I understand if this is still deemed not worth the complexity increase, but I wanted to point this out as further evidence that this would be a highly reusable abstraction.  I appreciate the desire to keep .talon files simple, and I appreciate that some things simply belong in python in captures. But I do think if there's a way to use talon files to support the subset of captures that result in _actions_, that aligns with what talon files already do well, and it would be really powerful and lead to cleaner/simpler code in cases like this.

**Comment by lunixbochs** [comment]
Exposing a whole mode as a single rule is just begging for accidental quadratic graphs. Rephrase is a much simpler way to do that and already works

**Comment by wolfmanstout** [comment]
FWIW, if I try to use prose and halt in one utterance (e.g. "prose this is a test halt tab new") I get an error:
```
talon.scripting.talon_script.TalonScriptError: 
 in script at C:\Users\james\AppData\Roaming\talon\user\knausj_talon\modes\dictation_mode.talon:33:
   mode.disable('sleep')
   mode.enable('command')
   mode.disable('dictation')
 > user.parse_phrase(phrase or '')
RuntimeError: Cannot recursively call action: "user.parse_phrase"
```

Perhaps that's worthy of a separate bug, but my own preference is a little different from @pokey's: I'd like to have something like `speak <user.dictation_commands>+$`, so that speak is totally consistent with dictation mode (i.e. except effectively modeless).   On the other hand, if I wanted similar support for "sentence", I suppose this approach wouldn't work because I'd need to capture the text in order to capitalize it. Hmm, the more I think about this, the more likely I am to use a beefy capture that returns a string.

**Comment by lunixbochs** [comment]
https://github.com/talonvoice/talon/issues/94

**Comment by wolfmanstout** [comment]
I'd like to share where I landed for my dictation-related grammars, since that was the original driver for this bug and Ryan suggested this might be worth considering doing upstream in knausj. Ultimately, I found value *both* in using anchor-file in the way originally proposed in my first comment (to separate chaining of dictation commands from control commands) *and* in extending the user.prose capture to include words that alter formatting like "cap", "no caps", and "no space". It's helpful having this functionality in a capture that returns text so that it can be used in a wider variety of contexts than a capture that can only be executed to type into the current context (e.g. I use it to specify query text used to move the cursor). Although I _could_ then reuse this capture for dictation mode and avoid anchor-file, I find it's nice to have commands like "cap" and "no space" work standalone to modify the formatter state without having to speak them in the same utterance as the text that follows. That's why I also have these commands in a file with anchor-file.  Behind the scenes, I'm trying to reuse as much code as possible. In fact I managed to remove some duplication in dictation.py by using the DictationFormat object to back both the prose capture and the auto_insert commands.  For the commands, this is a global that maintains state (e.g. whether to capitalize the next word), just as it does today. For the captures, this is instantiated locally only for the duration of the capture. This gives me both modal and modeless commands at my disposal, with consistent behavior across each.

You can see all the code for this in my fork. Here's my modified dictation_mode.talon with anchor-file:
https://github.com/wolfmanstout/wolfmanstout_talon/blob/master/modes/dictation_mode.talon

Here's where I've included generic_editor.talon commands in dictation mode (but thanks to anchor-file they can't be interleaved with actual dictation):
https://github.com/wolfmanstout/wolfmanstout_talon/blob/master/text/generic_editor.talon

And here's the code that backs dictation_mode.talon and the user.prose capture:
https://github.com/wolfmanstout/wolfmanstout_talon/blob/master/code/dictation.py

There are some other improvements here, e.g. I figured out a way to get spacing to work correctly around quotes, and I've added some "prose_snippets" that function similarly to `{user.punctuation}` but are not restricted to single keypresses.  The latter could potentially be implemented as part of the vocabulary file, but I've included them here for control over whitespace and because they seem more central than the custom vocabulary.

Would be curious to hear from others if they think this system would work for them. I'm still making minor tweaks here and there, but overall I'm pretty happy with the UX.

**Comment by lunixbochs** [comment]
this is done (without anchor-file) in the latest beta. you can enable both command and dictation mode in unmodified knausj_talon and it will anchor them separately.

note I want to test the implications of this for longer before fully committing to it, so I don't think this change will make it into the upcoming 0.4 public release.
